File: src-tauri/plugins/ble-peripheral/android/src/main/java/com/phantomhr/plugin/HrDataGenerator.kt
package com.phantomhr.plugin

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlin.math.max
import kotlin.math.min
import kotlin.random.Random

/**
 * Generates realistic 1 Hz heart rate data using a smoothed random walk.
 *
 * Algorithm per tick:
 * - 70%: small step: random pick from {-1, 0, +1}
 * - 30%: larger step: random pick from {-2, -1, 0, +1, +2}
 * Result is clamped to [baseBpm - variance, baseBpm + variance].
 *
 * When [updateConfig] is called mid-session, currentBpm drifts toward the new
 * baseBpm naturally via the clamp — no abrupt jump.
 */
class HrDataGenerator(config: HrConfig) {

    @Volatile private var baseBpm: Int = config.baseBpm
    @Volatile private var variance: Int = config.variance

    // Not volatile — only accessed from the single coroutine launched in start()
    private var currentBpm: Int = config.baseBpm

    private var job: Job? = null

    private val _bpmFlow = MutableSharedFlow<Int>(replay = 1)
    val bpmFlow: SharedFlow<Int> = _bpmFlow.asSharedFlow()

    /** Begin emitting BPM values at 1 Hz on [scope]. Safe to call only once per instance. */
    fun start(scope: CoroutineScope) {
        currentBpm = baseBpm
        job = scope.launch {
            while (isActive) {
                delay(1000L)
                currentBpm = nextBpm(currentBpm)
                _bpmFlow.emit(currentBpm)
            }
        }
    }

    /** Cancel the emission coroutine. No-op if already stopped. */
    fun stop() {
        job?.cancel()
        job = null
    }

    /**
     * Update generation parameters. The current BPM drifts toward the new
     * [newConfig.baseBpm] via clamping; it does not jump immediately.
     */
    fun updateConfig(newConfig: HrConfig) {
        baseBpm = newConfig.baseBpm
        variance = newConfig.variance
    }

    private fun nextBpm(current: Int): Int {
        val step = if (Random.nextFloat() < 0.7f) {
            Random.nextInt(-1, 2)  // -1, 0, or +1
        } else {
            Random.nextInt(-2, 3)  // -2, -1, 0, +1, or +2
        }
        val lo = max(40, baseBpm - variance)
        val hi = min(200, baseBpm + variance)
        return (current + step).coerceIn(lo, hi)
    }
}

File: src-tauri/plugins/ble-peripheral/ios/Sources/HrDataGenerator.swift
import Foundation

/// Generates realistic 1 Hz heart rate data using a smoothed random walk.
///
/// Algorithm per tick:
/// - 70%: small step from {-1, 0, +1}
/// - 30%: larger step from {-2, -1, 0, +1, +2}
/// Result is clamped to [baseBpm - variance, baseBpm + variance].
///
/// When ``updateConfig(baseBpm:variance:)`` is called mid-session, currentBpm
/// drifts toward the new baseBpm naturally via the clamp — no abrupt jump.
///
/// All methods and timer callbacks run on the caller-provided serial DispatchQueue,
/// so no locking is needed.
class HrDataGenerator {

    private var baseBpm: Int
    private var variance: Int
    private var currentBpm: Int

    private var timer: DispatchSourceTimer?
    private let queue: DispatchQueue

    init(baseBpm: Int, variance: Int, queue: DispatchQueue) {
        self.baseBpm = baseBpm
        self.variance = variance
        self.currentBpm = baseBpm
        self.queue = queue
    }

    /// Begin emitting BPM values at 1 Hz. `onTick` fires on `queue`.
    func start(onTick: @escaping (Int) -> Void) {
        stop()
        currentBpm = baseBpm

        let src = DispatchSource.makeTimerSource(queue: queue)
        src.schedule(deadline: .now() + 1.0, repeating: 1.0)
        src.setEventHandler { [weak self] in
            guard let self = self else { return }
            self.currentBpm = self.nextBpm(self.currentBpm)
            onTick(self.currentBpm)
        }
        src.resume()
        timer = src
    }

    /// Cancel the timer. No-op if already stopped.
    func stop() {
        timer?.cancel()
        timer = nil
    }

    /// Update generation parameters. Current BPM drifts toward new base via clamping.
    func updateConfig(baseBpm: Int, variance: Int) {
        self.baseBpm = baseBpm
        self.variance = variance
    }

    // MARK: - Private

    private func nextBpm(_ current: Int) -> Int {
        let step: Int
        if Float.random(in: 0..<1) < 0.7 {
            step = Int.random(in: -1...1)
        } else {
            step = Int.random(in: -2...2)
        }
        let lo = max(40, baseBpm - variance)
        let hi = min(200, baseBpm + variance)
        return min(hi, max(lo, current + step))
    }
}
